### ISSUE TITLE
SC2230 - command -v is not a direct replacement for which

### ISSUE THREAD

**Issue by arth1** [koalaman/shellcheck#1162]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC2230
- My shellcheck version (`shellcheck --version` or "online"): 4.4.7 (latest from git)
- [X] I tried on shellcheck.net and verified that this is still a problem on the latest commit
#### Here's a snippet or screenshot that shows the problem:

```sh

#!/bin/sh

which "$0"
/usr/bin/which "$0"
command -v "$0"

```

#### Here's what shellcheck currently says:

In test.sh line 3:
which "$0"
^-- SC2230: which is non-standard. Use builtin 'command -v' instead.

In test.sh line 4:
/usr/bin/which "$0"
^-- SC2230: which is non-standard. Use builtin 'command -v' instead.

#### Here's what I wanted or expected to see:

Because "command -v" does not do the same as which, I would have liked to see nothing.
When running the above (from /tmp), I get:
/tmp/test.sh
/tmp/test.sh
./test.sh

While "which" may not be standard, it is commonly used to get the full path.
At the very least I would like to not get warned when I specifically use the path as in the third command - then it's fairly certain what the user wants, which won't be the common bash alias.

**Comment by Dmole** [comment]
You could use 'pwd'.

'which' is not posix
http://pubs.opengroup.org/onlinepubs/9699919799/idx/utilities.html
and not gnucore
https://en.m.wikipedia.org/wiki/List_of_GNU_Core_Utilities_commands

but it sure seems common to me.. where is it not? 
also shellcheck is not warning about far less common things like Java... seems like a false positive to me.

**Comment by brodygov** [comment]
`command -v` also has slightly different behavior since it will return success for shell built-ins and functions that do not exist as standalone commands, whereas `which` only searches the PATH.

For example:

```sh
$ which if
# exits nonzero

$ command -v if
if
# exits zero

$ foo() { true ; }
$ which foo
# exits nonzero
$ command -v foo
foo
# exits zero
```

**Comment by pedroarthur** [comment]
Besides that, `which a b c` exits zero if __all__ of `a`, `b`, and `c` exist. Meanwhile, `command -v a b c` exits zero if __any__ of `a`, `b`, or `c` exists.

**Comment by pdietl** [comment]
So do we think that this check should be removed?

**Comment by brodygov** [comment]
:+1: I would favor removal.

**Comment by leagris** [comment]
An alternative to `which command`, when you need to know the path of a non built-in command:
```Bash
type -P ls
/bin/ls
```

**Comment by oliv3r** [comment]
While which can certainly be used if a user decides to (they can easily [ignore](https://github.com/koalaman/shellcheck/wiki/Ignore) the shellcheck warning, the argument 'but it is so common and always available' is of course moot and opinionated (yes, even busybox features it these days).

The point remains, if you use /bin/sh you kind of imply POSIX compliance, and even if not, some of us would still like this warning to be triggered. Just as which not being standard (again, even with being so ever common).

As with everything, there are exceptions of course. But I think this is not one of them, the exception should be in your code or your parser, where you globally or locally ignore this test as that would work for you. (You being not the individual personal you).

**Comment by pedroarthur** [comment]
This is a example where `which` and `command` diverge. Using `which`, I check if **all** the tools I need exist with a single statement. If I use, `command -V`, it will return 0 if **any** of the tools exist.

```bash
#!/bin/bash
tooling=( bash no_such_command )

# shellcheck disable=2230
which      "${tooling[@]}" &> /dev/null || echo some tools are missing
command -v "${tooling[@]}" &> /dev/null || echo some tools are missing
```

Hence, `command -v` is not a replacement for `which`.

**Comment by oliv3r** [comment]
is this even supposed to work?
I use it as such:
```sh
check_requirements()
{
    for cmd in ${CMDS}; do
        if ! test_result="$(command -V "${cmd}")"; then
            test_result_fail="${test_result_fail:-}${test_result}\n"
        else
            test_result_pass="${test_result_pass:-}${test_result}\n"
        fi
    done

    echo "Passed tests:"
    # As the results contain \n, we expect these to be interpreted.
    # shellcheck disable=SC2059
    printf "${test_result_pass:-none\n}"
    echo
    echo "Failed tests:"
    # shellcheck disable=SC2059
    printf "${test_result_fail:-none\n}"
    echo

    if [ -n "${test_result_fail:-}" ]; then
        echo "Self-test failed, missing dependencies."
        exit 1
    fi
}
```
which yes, is a lot longer, but also is a _lot_ friendlier to the user imo :)

**Comment by pedroarthur** [comment]
> is this even supposed to work?

Yes, `which` is supposed to work this way since day 0. Have you read its man page?

> so yeah, which can do this in one line, but just because you can, doesn't mean you should imo :)

So, you advocate to maintain 15+ LoC that triggers far more concerning checks instead of a single one that just has a "non-standard tool" warning? Thank you, but no.

However, the main issue here is saying that things are interchangeable when they aren't. Following the advice in this check can break perfectly working code.

**Comment by oliv3r** [comment]
> > is this even supposed to work?
> 
> Yes, `which` is supposed to work this way since day 0. Have you read its man page?
Nope :)
> 
> > so yeah, which can do this in one line, but just because you can, doesn't mean you should imo :)
> 
> So, you advocate to maintain 15+ LoC that triggers far more concerning checks instead of a single one that just has a "non-standard tool" warning? Thank you, but no.
more concerning checks? I personally prefer to have an output like:

Fair point, which does the same in one go, which is indeed (no pun intended) admittedly quite nice. So at the least, the wording could be improved absolutely. But the fact remains, and the warning, that it is a non-standard tool. So a green warning should be sufficient.

> 
> However, the main issue here is saying that things are interchangeable when they aren't. Following the advice in this check can break perfectly working code.

**Comment by HalisCz** [comment]
I've found out another difference between `which` and `command -v`:
```
$ alias ls
alias ls='ls -al'
$ which ls
/usr/bin/ls
$ command -v ls
alias ls='ls -al'
$ type -P ls
/usr/bin/ls
```

Therefore I think the suggestion should be _replace `which` with `type P`_.

**Comment by howaboutsynergy** [comment]
The problem is that `type -P` doesn't work in some shells that are used as `/bin/sh`, such as `dash`. [info](https://stackoverflow.com/questions/2608688/check-if-a-program-exists-in-bash/2609292#2609292)

For a `which` (function) that works on bash(/sh), dash and zsh, by using `type`(optionally with `-P` if supported) see [this gist](https://gist.github.com/constantoverride/591d65c093bfec522d4b22b99bf2c24b)

PR that added the `command -v` suggestion: https://github.com/koalaman/shellcheck/pull/1123
(linking here so that PR will have a link back to this issue)

**Comment by leagris** [comment]
Bash Implementation of `which`

```Bash
function which() {
  local IFS=:
  local f
  local p
  local -a pa
  read -r -d '' -a pa < <(printf '%s:\0' "${PATH}")
  for p in "${pa[@]}"; do
    f="${p}/${1}"
    if [[ -L "${f}" ]]; then
      f="$(readlink "${i}/${1}")"
    fi
    if [[ -f "${f}" && -x "${f}" ]]; then
      echo "${p}/${1}"
      return
    fi
  done
  return 1
}
```

**Comment by abathur** [comment]
I just lost about 40 minutes debugging this one, assuming I'd introduced the error alongside quoting fixes around some alias commands that I made at the same time.

I'm still vaguely glad the issue came up--I realized I could at least swap which for type -P and shave a little execution time off--but it would be nice if the check language directly acknowledged that these aren't fungible replacements. 

The current language is quite confident, and I naively took it at face value.

**Comment by tripleee** [comment]
The fundamental problem with `which` is that it's not reliable. There is a plethora of incompatible implementations, none of which are well-defined or portable. The recommendation should certainly be tweaked to not imply that `command -v` is always what you want to replace `which` with; in so many words, we could say "try `command -v` or `type` (or perhaps `type -p` if you are on Bash)".

Here's an excellent background on the topic by the always virtuose Stéphane Chazelas:
https://unix.stackexchange.com/a/85250/19240

**Comment by Dmole** [comment]
Shellcheck requires the shell to be specified with #! so non-standard shells can be treated independently of sh/bash.

**Comment by AdamDanischewski** [comment]
Shellcheck's [web page](https://www.shellcheck.net/), states: 
> ShellCheck
> finds **bugs** in your shell scripts.

It says that it finds **_bugs_** it doesn't say anything about portability.

If finding bugs is the goal, then shellcheck should keep quiet about people on Bash using `which`, `type -P`, etc - since it is not an error. 

Shell check probably needs to beef up its logic to handle differences between major shells - for instance `type -P` exists reliably in Bash but not in Dash. 

Also, I didn't see anyway to specify the shell version - I suppose we should presume a recent version but it would be nice if shellcheck displayed what version of a shell it was running against.

**Comment by Dmole** [comment]
> 2. It might be nice to have:

In vim shellcheck lists Errors, and Warnings.
info things use the warning tag but start with "Note that".
 examples;
```
error| Don't use $ on the left side of assignments. [SC1066]
warning| X appears unused. Verify it or export it. [SC2034]
warning| Note that, unescaped, this expands on the client side. [SC2029]
```

**Comment by beltaurus** [comment]
Indeed! I completely blanked here...
They are even color-coded in Visual Studio Coder:
![image](https://user-images.githubusercontent.com/54114043/66223490-77b80480-e688-11e9-9eba-9ce28757826c.png)

That's what I get for commenting before my morning coffee...

**Comment by AdamDanischewski** [comment]
Maybe performance should also be considered: 
![Screenshot from 2019-10-06 08-46-10](https://user-images.githubusercontent.com/10157303/66269390-d19a0500-e815-11e9-9a36-2636d03fa44c.png)

**Comment by abathur** [comment]
Apparently this check defaults to off since https://github.com/koalaman/shellcheck/commit/0f15fa49ba8cc427aaad338140f82707b6865eb5 was released in 0.7.1.

**Comment by EricLuehrsen** [comment]
> 
>     2. It might be nice to have:
>        "Errors" (bugs)
>        "Warnings" (portability issues)
>        "Info" (coding style).

I tried suggesting a little more lint-like categories a while ago. It would be nice if the numbering scheme followed something like error (1000), warning (2000), info (3000), portability (4000), and best-practice/stye (5000). Deploying shell check as quality tool could be done as many lint-like tools. Over time it would easier to ratchet QA checks on system scripts and supported tools. Both of these ideas were not accepted.

Also `/usr/bin/which` and unspecific `which` should have different enable options. The former enabled for "pedantic POSIX" option, and the later normally enabled for naive system administrators. They could even be the same message number but one is "/usr/bin/which _may not be_ portable ..." and the other "which _is not_ portable ..." This is done for a number of GCC diagnostics.

**Comment by scop** [comment]
I wonder if there's a reason not to recommend `type -P` _where it's known to work_, both in shellcheck output and at least in wiki? It does work at least with bash, seemingly also with ksh. It was mentioned a couple of times here, the only drawback brought up being portability.

Instead, there are mentions about `type -p` when on bash in wiki. But given the availability of `type -P` in bash, I don't know why one would ever want to use `type -p` in it as a `which` replacement.

**Comment by escoand** [comment]
I've found a somehow portabel solution working on IBM AIX, msys2 and RHEL7.

```sh
$ alias grep=true
$ command -v grep
alias grep=true

$ which grep # on IBM AIX
/opt/freeware/bin/grep
$ which grep # on RHEL7
alias grep='true'
        /usr/bin/true 

$ whence grep # on IBM AIX
true
$ whence grep # on RHEL7 and msys2
-bash: whence: command not found

$ type -P grep
/usr/bin/grep
$ type -P grep # on IBM AIX
ksh: whence: 0403-010 A specified flag is not valid for this command.

$ sh -c "command -v grep" # working on all
```
